sdk/src/main/java/io/opentelemetry/sdk/metrics/BoundRegistry.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics;
+
+import io.opentelemetry.metrics.LabelSet;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+abstract class BoundRegistry<B extends AbstractBoundInstrument> {
+  private final ConcurrentHashMap<LabelSet, B> bounds;
+  private final ReentrantLock collectLock;
+
+  BoundRegistry() {
+    bounds = new ConcurrentHashMap<>();
+    collectLock = new ReentrantLock();
+  }
+
+  B acquire(LabelSet labelSet) {
+    B bound = bounds.get(labelSet);
+    if (bound != null && bound.ref()) {
+      // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+      return bound;
+    }
+
+    // Missing entry or no longer mapped, try to add a new entry.
+    bound = newBound(labelSet);
+    while (true) {
+      B oldBound = bounds.putIfAbsent(labelSet, bound);
+      if (oldBound != null) {
+        if (oldBound.ref()) {
```

Comment:
rather than using ref counting, can we instead use a map with weak references for values? This ref counting seems like an overly baroque way to deal with this issue, and adds a significant amount of complexity to this code.

Reply:
I hope now it is clearer why I used this, we have threads on critical path adding bounds, then the collector needs to collect them then remove them from the bounds, using weak references may cause uncollected data.